File: src/main/native/cpp/Subsystems/Swerve/SwerveChassis/SwerveChassis.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.
#include "OvertureLib/Subsystems/Swerve/SwerveChassis/SwerveChassis.h"
#include <networktables/NetworkTableInstance.h>
#include <networktables/StructTopic.h>

/**
 * @brief Builds an object of swerve chassis
 */
SwerveChassis::SwerveChassis() : SwerveBase(this) {
	auto config = pathplanner::RobotConfig::fromGUISettings();

	m_setpointGenerator = pathplanner::SwerveSetpointGenerator(config,
			12.5_tps);

	frc::ChassisSpeeds speeds = getCurrentSpeeds();
	std::vector < frc::SwerveModuleState > currentStates(4);
	for (int i = 0; i < 4; i++) {
		currentStates[i] = frc::SwerveModuleState();
	}
	previousSetpoint = pathplanner::SwerveSetpoint(speeds, currentStates,
			pathplanner::DriveFeedforwards::zeros(config.numModules));
}

/**
 * @brief Disbable the speed helper
 */
void SwerveChassis::disableSpeedHelper() {
	speedsHelper = std::nullopt;
}

/**
 * @brief Sets the speed helper
 *
 * @param headingOverride The heading override
 */
void SwerveChassis::enableSpeedHelper(SpeedsHelper *speedsHelper) {
	speedsHelper->initialize();
	this->speedsHelper = speedsHelper;
}

/**
 * @brief Returns the robot relative speeds
 *
 * @return ChassisSpeeds object
 */
frc::ChassisSpeeds SwerveChassis::getCurrentSpeeds() {
	return currentSpeeds;
}

/**
 * @brief Returns the robot accelerations
 *
 * @return ChassisAccels object
 */
ChassisAccels SwerveChassis::getCurrentAccels() {
	return currentAccels;
}

/**
 * @brief Returns the robot odometry
 *
 * @return Pose2d object
 */
const frc::Pose2d& SwerveChassis::getEstimatedPose() {
	return latestPose;
}

/**
 * @brief Resets the robot odometry
 *
 * @param initPose Pose2d object
 */
void SwerveChassis::resetOdometry(frc::Pose2d initPose) {
	odometry->ResetPosition(getRotation2d(), modulesPositions, initPose);
}

/**
 * @brief Updates odometry using vision
 */
void SwerveChassis::addVisionMeasurement(frc::Pose2d pose,
		units::second_t timestamp) {
	if (!acceptingVisionMeasurements) {
		return;
	}
	odometry->AddVisionMeasurement(pose, timestamp);
}

void SwerveChassis::setAcceptingVisionMeasurements(
		bool acceptingVisionMeasurements) {
	this->acceptingVisionMeasurements = acceptingVisionMeasurements;
}

/**
 * @brief Sets the odometry to desired angle
 *
 * @param angle Desired angle
 */
void SwerveChassis::resetHeading(units::degree_t angle) {
	frc::Pose2d actualOdometry = getEstimatedPose();
	frc::Pose2d newOdometry { actualOdometry.X(), actualOdometry.Y(), angle };
	resetOdometry(newOdometry);
}

void SwerveChassis::setTargetSpeeds(frc::ChassisSpeeds speeds) {
	desiredSpeeds = speeds;
}

void SwerveChassis::setModuleStates(
		const std::vector<frc::SwerveModuleState> &desiredStates) {
	getFrontLeftModule().setState(desiredStates[0]);
	getFrontRightModule().setState(desiredStates[1]);
	getBackLeftModule().setState(desiredStates[2]);
	getBackRightModule().setState(desiredStates[3]);
}
/**
 * @brief Runs the SysId Quasisstatic command
 */
frc2::CommandPtr SwerveChassis::SysIdQuadstatic(
		frc2::sysid::Direction direction) {
	return frc2::cmd::Sequence(frc2::InstantCommand([this]() {
		characterizing = true;
		sysIdVoltage(0_V);
	}).ToPtr(), frc2::cmd::Wait(0.5_s), m_sysIdRoutine.Quasistatic(direction)).FinallyDo(
			[this] {
				characterizing = false;
			});
}

/**
 * @brief Runs the SysId Dynamic command
 */
frc2::CommandPtr SwerveChassis::SysIdDinamic(frc2::sysid::Direction direction) {
	return frc2::cmd::Sequence(frc2::InstantCommand([this]() {
		characterizing = true;
		sysIdVoltage(0_V);
	}).ToPtr(), frc2::cmd::Wait(0.5_s), m_sysIdRoutine.Dynamic(direction)).FinallyDo(
			[this] {
				characterizing = false;
			});
}

/**
 * @brief Sets the voltage for the SysId command
 */
void SwerveChassis::sysIdVoltage(units::volt_t voltage) {
	getFrontLeftModule().setVoltageDrive(voltage);
	getFrontRightModule().setVoltageDrive(voltage);
	getBackRightModule().setVoltageDrive(voltage);
	getBackLeftModule().setVoltageDrive(voltage);
}

/**
 * @brief Updates the robot odometry
 */
void SwerveChassis::updateOdometry() {
	odometry->Update(getRotation2d(), modulesPositions);
	latestPose = odometry->GetEstimatedPosition();
	currentSpeeds = getKinematics().ToChassisSpeeds(modulesStates);

	currentAccels = ChassisAccels(currentSpeeds, lastSpeeds);
	lastSpeeds = currentSpeeds;
}

void SwerveChassis::shuffleboardPeriodic() {
	frc::SmartDashboard::PutNumber("SwerveChassis/Odometry/LinearX",
			desiredSpeeds.vx.value());
	frc::SmartDashboard::PutNumber("SwerveChassis/Odometry/LinearY",
			desiredSpeeds.vy.value());
	frc::SmartDashboard::PutNumber("SwerveChassis/Odometry/Angular",
			desiredSpeeds.omega.value());

	frc::SmartDashboard::PutNumber("SwerveChassis/Odometry/AccelX",
			currentAccels.ax.value());
	frc::SmartDashboard::PutNumber("SwerveChassis/Odometry/AccelY",
			currentAccels.ay.value());
	frc::SmartDashboard::PutNumber("SwerveChassis/Odometry/AccelOmega",
			currentAccels.omega.value());

	frc::SmartDashboard::PutNumber("SwerveChassis/Odometry/SpeedX",
			desiredSpeeds.vx.value());
	frc::SmartDashboard::PutNumber("SwerveChassis/Odometry/SpeedY",
			desiredSpeeds.vy.value());
	frc::SmartDashboard::PutNumber("SwerveChassis/Odometry/SpeedOmega",
			desiredSpeeds.omega.value());

	frc::SmartDashboard::PutNumber("SwerveChassis/Odometry/AcceptingVision", acceptingVisionMeasurements);

	posePublisher.Set(latestPose);

	frc::SmartDashboard::PutNumberArray("SwerveChassis/Control/DesiredSpeeds",
			std::vector { desiredSpeeds.vx.value(), desiredSpeeds.vy.value(),
					desiredSpeeds.omega.value() });

	getFrontLeftModule().shuffleboardPeriodic();
	getFrontRightModule().shuffleboardPeriodic();
	getBackRightModule().shuffleboardPeriodic();
	getBackLeftModule().shuffleboardPeriodic();
}

void SwerveChassis::Periodic() {
	if (!configuredChassis) {
		throw new std::runtime_error(
				"Have not called SwerveBase::configureSwerveBase!!!");
	}

	if (characterizing) {
		return;
	}

	if (speedsHelper) {
		speedsHelper.value()->alterSpeed(desiredSpeeds);
	}

	modulesPositions[0] = getFrontLeftModule().getPosition();
	modulesPositions[1] = getFrontRightModule().getPosition();
	modulesPositions[2] = getBackLeftModule().getPosition();
	modulesPositions[3] = getBackRightModule().getPosition();

	modulesStates[0] = getFrontLeftModule().getState();
	modulesStates[1] = getFrontRightModule().getState();
	modulesStates[2] = getBackLeftModule().getState();
	modulesStates[3] = getBackRightModule().getState();

	frc::ChassisSpeeds targetSpeeds = { getVxLimiter().Calculate(
			desiredSpeeds.vx), getVyLimiter().Calculate(desiredSpeeds.vy),
			getVwLimiter().Calculate(desiredSpeeds.omega) };

	wpi::array < frc::SwerveModuleState, 4U > desiredStates =
			getKinematics().ToSwerveModuleStates(targetSpeeds);
	getKinematics().DesaturateWheelSpeeds(&desiredStates, getMaxModuleSpeed());

	std::vector < frc::SwerveModuleState
			> desiredStatesVector(desiredStates.begin(), desiredStates.end());

	updateOdometry();
	poseLog.Append(latestPose);

	setModuleStates (desiredStatesVector);
}

File: src/main/java/com/overture/lib/subsystems/Swerve/SwerveChassis.java
package com.overture.lib.subsystems.Swerve;

import static edu.wpi.first.units.Units.Meters;
import static edu.wpi.first.units.Units.MetersPerSecond;
import static edu.wpi.first.units.Units.RotationsPerSecond;
import static edu.wpi.first.units.Units.Seconds;
import static edu.wpi.first.units.Units.Volts;

import com.overture.lib.math.ChassisAccels;
import com.overture.lib.subsystems.Swerve.SpeedsHelper.SpeedsHelper;
import com.pathplanner.lib.auto.AutoBuilder;
import com.pathplanner.lib.config.PIDConstants;
import com.pathplanner.lib.controllers.PPHolonomicDriveController;
import com.pathplanner.lib.util.DriveFeedforwards;
import com.pathplanner.lib.util.swerve.SwerveSetpoint;
import com.pathplanner.lib.util.swerve.SwerveSetpointGenerator;
import edu.wpi.first.math.estimator.SwerveDrivePoseEstimator;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.math.kinematics.SwerveModulePosition;
import edu.wpi.first.math.kinematics.SwerveModuleState;
import edu.wpi.first.units.measure.Voltage;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.smartdashboard.Field2d;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.Commands;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import edu.wpi.first.wpilibj2.command.sysid.SysIdRoutine;
import edu.wpi.first.wpilibj2.command.sysid.SysIdRoutine.Direction;
import java.util.Optional;

/** The SwerveChassis class */
public class SwerveChassis extends SubsystemBase {
  private SwerveSetpointGenerator m_setpointGenerator;
  private SwerveSetpoint previousSetpoint;
  private Pose2d latestPose;

  private ChassisSpeeds desiredSpeeds = new ChassisSpeeds();
  private ChassisSpeeds currentSpeeds = new ChassisSpeeds();
  private ChassisSpeeds lastSpeeds = new ChassisSpeeds();
  private ChassisAccels currentAccels = new ChassisAccels();

  private Field2d field2d;

  private SwerveDrivePoseEstimator odometry;
  private SwerveModulePosition[] modulePositions = new SwerveModulePosition[4];
  private SwerveModuleState[] moduleStates = new SwerveModuleState[4];

  Optional<SpeedsHelper> speedsHelper;
  boolean acceptingVisionMeasurements = false;
  boolean characterizing = false;

  SwerveConfig config;
  SysIdRoutine m_sysIdRoutine;

  /**
   * Constructor for SwerveChassis
   *
   * @param config - SwerveConfig
   */
  public SwerveChassis(SwerveConfig config) {
    this.config = config;

    // Configure AutoBuilder last
    AutoBuilder.configure(
        this::getEstimatedPose,
        this::resetOdometry,
        this::getCurrentSpeeds,
        (speeds, feedforwards) -> setTargetSpeeds(speeds),
        new PPHolonomicDriveController(
            new PIDConstants(5.0, 0.0, 0.0), new PIDConstants(5.0, 0.0, 0.0)),
        config.pathplannerConfig,
        () -> {
          var alliance = DriverStation.getAlliance();
          if (alliance.isPresent()) {
            return alliance.get() == DriverStation.Alliance.Red;
          }
          return false;
        },
        this);

    odometry =
        new SwerveDrivePoseEstimator(
            config.kinematics, new Rotation2d(), modulePositions, new Pose2d());

    SmartDashboard.putData("Field", field2d);

    m_sysIdRoutine =
        new SysIdRoutine(
            new SysIdRoutine.Config(),
            new SysIdRoutine.Mechanism(
                this::sysIdVoltage,
                log -> {
                  log.motor("frontLeft")
                      .voltage(config.swerveModules[0].getDriveVoltage())
                      .linearPosition(
                          Meters.of(config.swerveModules[0].getPosition().distanceMeters))
                      .linearVelocity(
                          MetersPerSecond.of(
                              config.swerveModules[0].getState().speedMetersPerSecond));
                  log.motor("frontRight")
                      .voltage(config.swerveModules[1].getDriveVoltage())
                      .linearPosition(
                          Meters.of(config.swerveModules[1].getPosition().distanceMeters))
                      .linearVelocity(
                          MetersPerSecond.of(
                              config.swerveModules[1].getState().speedMetersPerSecond));
                  log.motor("backRight")
                      .voltage(config.swerveModules[2].getDriveVoltage())
                      .linearPosition(
                          Meters.of(config.swerveModules[2].getPosition().distanceMeters))
                      .linearVelocity(
                          MetersPerSecond.of(
                              config.swerveModules[2].getState().speedMetersPerSecond));
                  log.motor("backLeft")
                      .voltage(config.swerveModules[3].getDriveVoltage())
                      .linearPosition(
                          Meters.of(config.swerveModules[3].getPosition().distanceMeters))
                      .linearVelocity(
                          MetersPerSecond.of(
                              config.swerveModules[3].getState().speedMetersPerSecond));
                },
                this));

    m_setpointGenerator =
        new SwerveSetpointGenerator(config.pathplannerConfig, RotationsPerSecond.of(12.5));

    SwerveModuleState[] currentStates = new SwerveModuleState[4];
    previousSetpoint =
        new SwerveSetpoint(
            currentSpeeds,
            currentStates,
            DriveFeedforwards.zeros(config.pathplannerConfig.numModules));
  }

  /** Disables the speed helper */
  public void disableSpeedHelper() {
    speedsHelper = Optional.empty();
  }

  /**
   * Sets the speed helper
   *
   * @param speedsHelper SpeedsHelper object
   */
  public void enableSpeedHelper(SpeedsHelper speedsHelper) {
    speedsHelper.initialize();
    this.speedsHelper = Optional.of(speedsHelper);
  }

  /**
   * Returns the robot relative speeds
   *
   * @return ChassisSpeeds object
   */
  public ChassisSpeeds getCurrentSpeeds() {
    return currentSpeeds;
  }

  /**
   * Returns the robot accelerations
   *
   * @return ChassisAccels object
   */
  public ChassisAccels getCurrentAccels() {
    return currentAccels;
  }

  /**
   * Returns the robot odometry
   *
   * @return Pose2d object
   */
  public Pose2d getEstimatedPose() {
    return latestPose;
  }

  /**
   * Resets the robot odometry
   *
   * @param initPose Pose2d object
   */
  public void resetOdometry(Pose2d initPose) {
    odometry.resetPosition(config.pigeon.getRotation2d(), modulePositions, initPose);
  }

  /**
   * Updates odometry using vision
   *
   * @param pose Pose2d object
   * @param timestamp double
   */
  public void addVisionMeasurement(Pose2d pose, double timestamp) {
    if (acceptingVisionMeasurements) {
      odometry.addVisionMeasurement(pose, timestamp);
    }
  }

  /**
   * Sets if the robot is accepting vision measurements
   *
   * @param acceptingVisionMeasurements boolean
   */
  public void setAcceptingVisionMeasurements(boolean acceptingVisionMeasurements) {
    this.acceptingVisionMeasurements = acceptingVisionMeasurements;
  }

  /**
   * Sets the odometry to desired angle
   *
   * @param angle Desired angle
   */
  void resetHeading(Rotation2d angle) {
    resetOdometry(new Pose2d(getEstimatedPose().getTranslation(), angle));
  }

  /**
   * Sets the target speeds
   *
   * @param speeds ChassisSpeeds object
   */
  public void setTargetSpeeds(ChassisSpeeds speeds) {
    desiredSpeeds = speeds;
  }

  /**
   * Sets the target states
   *
   * @param desiredStates SwerveModuleState array
   */
  public void setModuleStates(SwerveModuleState[] desiredStates) {
    config.swerveModules[0].setState(desiredStates[0]);
    config.swerveModules[1].setState(desiredStates[1]);
    config.swerveModules[2].setState(desiredStates[2]);
    config.swerveModules[3].setState(desiredStates[3]);
  }

  /** Runs the SysId Quasisstatic command */
  Command SysIdQuadstatic(Direction direction) {
    return Commands.sequence(
            Commands.runOnce(
                () -> {
                  characterizing = true;
                  sysIdVoltage(Volts.of(0.0));
                }),
            Commands.waitTime(Seconds.of(0.5)),
            m_sysIdRoutine.quasistatic(direction))
        .finallyDo(
            () -> {
              characterizing = false;
            });
  }

  /**
   * Runs the SysId Dynamic command
   *
   * @param direction Direction object
   */
  Command SysIdDinamic(Direction direction) {
    return Commands.sequence(
            Commands.runOnce(
                () -> {
                  characterizing = true;
                  sysIdVoltage(Volts.of(0.0));
                }),
            Commands.waitTime(Seconds.of(0.5)),
            m_sysIdRoutine.dynamic(direction))
        .finallyDo(
            () -> {
              characterizing = false;
            });
  }

  /**
   * Sets the voltage for the SysId command
   *
   * @param voltage Voltage object
   */
  public void sysIdVoltage(Voltage voltage) {
    config.swerveModules[0].setVoltageDrive(voltage);
    config.swerveModules[1].setVoltageDrive(voltage);
    config.swerveModules[2].setVoltageDrive(voltage);
    config.swerveModules[3].setVoltageDrive(voltage);
  }

  /** Updates the robot odometry */
  public void updateOdometry() {
    odometry.update(config.pigeon.getRotation2d(), modulePositions);
    latestPose = odometry.getEstimatedPosition();
    currentSpeeds = config.kinematics.toChassisSpeeds(moduleStates);

    currentAccels = new ChassisAccels(currentSpeeds, lastSpeeds);
    lastSpeeds = currentSpeeds;
  }

  /** Shuffleboard update */
  public void shuffleboardPeriodic() {
    SmartDashboard.putNumber("Odometry/AccelX", currentAccels.ax.magnitude());
    SmartDashboard.putNumber("Odometry/AccelY", currentAccels.ay.magnitude());
    SmartDashboard.putNumber("Odometry/AccelOmega", currentAccels.omega.magnitude());

    SmartDashboard.putNumber("Odometry/SpeedX", currentSpeeds.vxMetersPerSecond);
    SmartDashboard.putNumber("Odometry/SpeedY", currentSpeeds.vyMetersPerSecond);
    SmartDashboard.putNumber("Odometry/SpeedOmega", currentSpeeds.omegaRadiansPerSecond);

    field2d.setRobotPose(latestPose);

    config.swerveModules[0].shuffleboardPeriodic();
    config.swerveModules[1].shuffleboardPeriodic();
    config.swerveModules[2].shuffleboardPeriodic();
    config.swerveModules[3].shuffleboardPeriodic();
  }

  /** Periodic function */
  @Override
  public void periodic() {
    if (characterizing) {
      return;
    }

    if (speedsHelper.isPresent()) {
      speedsHelper.get().alterSpeed(desiredSpeeds);
    }

    modulePositions[0] = config.swerveModules[0].getPosition();
    modulePositions[1] = config.swerveModules[1].getPosition();
    modulePositions[2] = config.swerveModules[2].getPosition();
    modulePositions[3] = config.swerveModules[3].getPosition();

    moduleStates[0] = config.swerveModules[0].getState();
    moduleStates[1] = config.swerveModules[1].getState();
    moduleStates[2] = config.swerveModules[2].getState();
    moduleStates[3] = config.swerveModules[3].getState();

    previousSetpoint =
        m_setpointGenerator.generateSetpoint(previousSetpoint, desiredSpeeds, Seconds.of(0.02));

    updateOdometry();

    setModuleStates(previousSetpoint.moduleStates());
  }
}
